src/unaudited/EcdsaEpochRegistry_Permissionless.sol

```diff
@@ -0,0 +1,250 @@
+    function _registerOperator(
+        address operator,
+        ISignatureUtils.SignatureWithSaltAndExpiry memory operatorSignature
+    ) internal virtual override {
+        if (operatorStatus[operator] != OperatorStatus.REGISTERED) {
+            require(weightOfOperator(operator) >= minimumWeightRequirement,
+                "insufficient weight to register");
+            operatorStatus[operator] = OperatorStatus.REGISTERED;
+
+            // Register the operator with the EigenLayer via this AVS's ServiceManager
+            serviceManager.registerOperatorToAVS(operator, operatorSignature);
+
+            // TODO: event
+            // emit OperatorRegistered(operator, operatorId);
+
+            operatorIndex[operator] = numberRegisteredOperators();
+            registeredOperators.push(operator);
+        }
+    }
+
+    // TODO: documentation
+    function _deregisterOperator(
+        address operator
+    ) internal virtual override {
+        if (operatorStatus[operator] == OperatorStatus.REGISTERED) {
+
+            operatorStatus[operator] = OperatorStatus.DEREGISTERED;
+            serviceManager.deregisterOperatorFromAVS(operator);
+
+            // TODO: event
+            // emit OperatorDeregistered(operator, operatorId);
+
+            // swap-and-pop routine to keep indices correct
+            uint256 numRegisteredOperators = numberRegisteredOperators();
+            uint256 deregisteredIndex = operatorIndex[operator];
+            address lastOperatorInArray = registeredOperators[numRegisteredOperators - 1];
+            registeredOperators[deregisteredIndex] = lastOperatorInArray;
+            operatorIndex[lastOperatorInArray] = deregisteredIndex;
+            delete operatorIndex[operator];
+            registeredOperators.pop();
+        }
+    }
+
+    // TODO: documentation
+    // only called if `currentEpoch() != 0`
+    function _adjustMinimumWeight(uint256 previousEpoch) internal virtual {
+        uint256 currentOperatorSetSize = numberRegisteredOperators();
+        uint256 newWeightRequirement;
+        if (currentOperatorSetSize == 0) {
+            newWeightRequirement = defaultWeightRequirement;
+        } else {
+            // store values in memory
+            uint256 previousWeightRequirement = minimumWeightRequirement;
+            uint256 _targetOperatorSetSize = targetOperatorSetSize;
+
+            // start with "adjustment" of 1 (i.e. leaving the `minimumWeightRequirement` value unchanged)
+            uint256 adjustmentFactor = 1e18;
+            if (currentOperatorSetSize >= _targetOperatorSetSize) {
+                // when current = target, no change
+                // when current = max, add the retargetingFactorWei
+                // when current is halfway between target and max, add half of retargetingFactorWei
+                adjustmentFactor += retargetingFactorWei * (currentOperatorSetSize - _targetOperatorSetSize) / (maxRegisteredOperators - _targetOperatorSetSize);
```

Comment:
The retargeting function is linear afaict

If this was based on the regular gas market EIP1559, it's a linear retargeting.  The blob space market 1559 is the only one that uses exponential retargeting, which has improved the smoothing of demand spikes and is the one we should base this off

Reply:
hmm, maybe I misunderstood the way these terms are being used.
I was viewing this as exponential retargeting since the net effect across epochs is the weight requirement being repeatedly multiplied by a factor (a la exponentiation), but I guess you're saying it would be exponential if the adjustment factor calculated was
`retargetingFactorWei` _raised to a power_ dependent on the difference between `currentOperatorSetSize` and `_targetOperatorSetSize`
instead of what is there now.
I think this is super reasonable; I specifically didn't try to optimize the algorithm here at all.

Reply:
makes sense, the latter seems more in line with what we want.  https://dankradfeist.de/ethereum/2022/03/16/exponential-eip1559.html